core/shared/src/main/scala/fs2/Stream.scala

```diff
@@ -15,7 +15,67 @@ import scala.concurrent.duration._
 
 /**
   * A stream producing output of type `O` and which may evaluate `F`
-  * effects. If `F` is [[Pure]], the stream evaluates no effects.
+  * effects.
+  * Or: a lazy sequence of computations that may evaluate `F` effects `F` and yield values of type `O`.
+  *
+  * Streams are purely functional: a `Stream[F, O]` is a description for a computation,
+  * much like `cats.effect.IO` or `scalaz.Task` are just descriptions of computations.
+  * The methods in the `Stream` class only calculate new descriptions from others.
+  *
+  * Streams are lazy, pull-based data sources. To '''evaluate''' a stream is to extract from it
+  * as many elements as a _consumer_ of the stream may need for the result.
+  *   by performing as many `F`-effectful computations as necessary to get those elements.
+  *   Thus, the effects in `F` are not evaluated unless and until they are required (pulled) by a stream consumer.
+  *
+  * - '''Laziness''': streams are lazily evaluated sources for effectful computations.
+  *   The operations on a Stream only evaluate a finite prefix,
+  *   a finite part of a whole stream, as needed to supply the requests of the stream _consumers_.
+  *   Thus, there can be infinite streams, or Streams based on
+  *   an unlimited source of information (e.g. a Signal that may be sampled with as frequently as desired).
+  *
+  * - '''Source''': a stream may simply be a fixed list of pre-defined F-effectful computations.
+  *   However, it can also be backed by an external data source (e.g. a network connection).
+  *   It can also be backed by a effectful computation that can be re-evaluated,
+  *   such as a Stream to read the current system time.
+  *
+  * There are some special properties or cases of streams:
+  *  - A '''pure''' stream is a `Stream` in which the `F` is [[Pure]], which evaluates no effects.
+  *  - A stream is '''terminating''', or '''finite''', if pulling from it reaches the end after yielding
+  *    a finite number of elements.
+  *  - In particular, a stream is ''''empty''' if it terminates after yielding no value.
+  *  - A '''singleton''' stream is a stream that ends after yielding one single value.
+  *  - A '''never''' stream is a stream that never terminates and never yields any value.
+  *
+  * == List-Like API and operations ==
+  *
+  * A `Stream[F, O]` can be (naively) seen as a list of elements with `F`-effects.
+  * Following the analogy between lists and streams, the API of the `Stream` class
+  * defines several methods similar to those of the `List` class.
+
+  * We can convert every ''pure and finite''' stream into a `List[O]` with the `.toList` method.
+  * This conversion obeys some equivalences between `Stream` and `List` operations.
+  *
+  *  - For any pure stream `s`, `s.foo().toList == s.toList.foo()`, is true if `foo` is one of `map`, `as`, `delete`, `take`, `drop`, `takeWhile`, `dropWhile`, `zipWithIndex`, `filter`, or `collect`.
```

Comment:
I'm not too keen on enumerating these methods here or in the table below. I think they will end up out of sync as new methods are added. Rather, I think this section can show some examples but shouldn't attempt to be exhaustive.

Reply:
Perhaps we could add one phrase on _Pure Semantics_ on each method's Scaladoc? The examples sort-of show that, but writing the equivalence would also be more informative.

Reply:
Yes, I will indicate that the absence of explicit constraints is a tacit description of purity.